pandas/_libs/index.pyx

```diff
@@ -1141,12 +1141,24 @@ cdef class ExtensionEngine(SharedEngine):
 
 cdef class MaskedIndexEngine(IndexEngine):
     def __init__(self, object values):
-        super().__init__(values._data)
-        self.mask = values._mask
+        super().__init__(self._get_data(values))
+        self.mask = self._get_mask(values)
+
+    def _get_data(self, object values) -> np.ndarray:
+        if hasattr(values, "_mask"):
+            return values._data
+        # We are an ArrowExtensionArray
+        return values.to_numpy(na_value=1, dtype=values.dtype.numpy_dtype)
```

Comment:
Curious why do we specify `na_value=1`?

Reply:
We have to use something that fits into all numeric arrays, otherwise we geht NA and end up with object

Reply:
Got it, could you add a comment to that effect here? With that and a rebase this LGTM

Reply:
Also a TODO to remove this once we have a separate indexing engine for pyarrow

Reply:
Done